### DISCUSSION TITLE
Keeping only the interior of bordered objects

### DISCUSSION THREAD

**Discussion by dakra137** [ImageMagick/ImageMagick#5379] [Help] [answered]
I want to "blank out" everything exterior to a closed polygon in each of many transparent PNG images produced by pygmt.  I can control the color of the polygon. 

The imagemagick documentation for doing this seems to be under construction or for adding borders. ref: [https://imagemagick.org/Usage/masking/#border_cut](https://imagemagick.org/Usage/masking/#border_cut)  and [https://legacy.imagemagick.org/Usage/crop/](https://legacy.imagemagick.org/Usage/crop/) 

Using the interactive imagemagick tool, this is easy to accomplish:

1. Open the file.
2. Select Image Edit // Matte 
3. Specify Matte Value == Transparent
4. Specify Border Color == Yellow
5. Specify Method == Fill to Border
6. Click on something non-transparent outside the border
7. Save the result

Any pointers or examples on how to do this from the command line, particularly the "Click on something non-transparent outside the border" part?

![Fig_1](https://user-images.githubusercontent.com/1835960/181205699-2a9b9f87-69a3-4ddc-a46d-3772609fb4a3.png)

**Comment by fmw42** [comment]
Threshold on the yellow color. Then floodfill the interior so that you have a filled polygon as white on a black background. Then invert that so that the polygon is black on a white background. Then put that image into the alpha channel of the original image.

**Comment by fmw42** [comment]
Imagemagick is command line. There is no click to select GUI. So you will have to provide that with some other tool and feed the coordinates to imagemmagick.

What is your IM version and platform/OS?

**Comment by fmw42** [comment]
This will extract the alpha channel. Then it will threshold on the yellow hexagon. Then negate it so that the hexagon is black on white background. Then it flood fills from 0,0 with black so that the hexagon is white filled on black background. This image is multiplied by the extracted alpha channel and is then put into the alpha channel of the input.

Assumes IM 6 Unix syntax:

```
convert yellow_hexagon.png -write mpr:img \
\( mpr:img -alpha extract \) \
\( mpr:img -alpha off -fuzz 10% -fill none +opaque yellow \
-fill white -opaque yellow -alpha off -negate \
-fuzz 0 -fill black -draw "color 0,0 floodfill" -alpha off  \) \
\( -clone 1,2 -compose multiply -composite \) \
-delete 1,2 \
-alpha off -compose copy_opacity -composite \
yellow_hexagon_result.png
```

![yellow_hexagon_result](https://user-images.githubusercontent.com/13458218/181403023-f039596f-387f-4c86-b390-0e61cb54f27f.png)

**Reply by dakra137** [reply]
The syntax from the original post worked for me, especially with the addition of two additional parameters  `-fill transparent  -transparent white` at the end to restore transparency to the interior.

```
convert Fig.1.png \( +clone -fuzz 10% -fill none +opaque yellow -fill white -opaque yellow \
    -alpha off -negate -fuzz 0 -fill black -draw "color 0,0 floodfill" -alpha off \) \
    -alpha off -compose copy_opacity -composite -fill transparent  -transparent white Fig.1k.png
```

As an aside, I would have thought that the following would have been sufficient, but it appears to accomplish no changes at all:
`convert Fig.1.png  -fill none  -fuzz 10%  -bordercolor yellow  -draw "color 1,1 filltoborder" Fig.1n.png
and the same for `
`convert Fig.1.png  -fill none  -fuzz 10%  -bordercolor yellow  -draw "matte 1,1 filltoborder" Fig.1n.png`

Thank you very much for your efforts and knowledge.  In every domain, it takes a lot of knowledge and imagination to create a long pathway of transformations to create something. 
David

**Reply by fmw42** [reply]
This should have been transparent already.  I did not notice the transparency in place of the white originally and edited my solution as below apparently after you have read my first draft of it.

So you can do in IM 6 either

```
convert yellow_hexagon.png -write mpr:img \
\( mpr:img -alpha extract \) \
\( mpr:img -alpha off -fuzz 10% -fill none +opaque yellow \
-fill white -opaque yellow -alpha off -negate \
-fuzz 0 -fill black -draw "color 0,0 floodfill" -alpha off  \) \
\( -clone 1,2 -compose multiply -composite \) \
-delete 1,2 \
-alpha off -compose copy_opacity -composite \
yellow_hexagon_result.png
```

or you can do in IM 6

```
convert yellow_hexagon.png -write mpr:img \
\( mpr:img -alpha extract \) \
\( mpr:img -fuzz 10% -fill none +opaque yellow \
-fill white -opaque yellow -alpha off  \
-fuzz 0 -fill white -draw "color 737,737 floodfill" -alpha off  \) \
\( -clone 1,2 -compose multiply -composite \) \
-delete 1,2 \
-alpha off -compose copy_opacity -composite \
yellow_hexagon_result2.png

```

where 737,737 is the center of the image (any point inside the yellow polygon).

**Reply by fmw42** [reply]
See edited code above.

**Comment by fmw42** [comment]
If you want the yellow included, then in IM 7, you can select the center automatically and do:

```
magick yellow_hexagon.png -write mpr:img \
\( mpr:img -alpha extract \) \
\( mpr:img -fuzz 10% -fill none +opaque yellow \
-fill white -opaque yellow -alpha off  \
-fuzz 0 -fill white -draw "color %[fx:w/2],%[fx:h/2] floodfill" -alpha off  \) \
\( -clone 1,2 -compose multiply -composite \) \
-delete 1,2 \
-alpha off -compose copy_opacity -composite \
yellow_hexagon_result2.png

```
![yellow_hexagon_result2](https://user-images.githubusercontent.com/13458218/181403228-58666b3f-740b-4a97-bbd3-7054e114cadb.png)

**Reply by dakra137** [reply]
I have to stick with your first solution in an earlier reply. My version is 6.9.11-60 Q16 x86_64 2021-01-25, which issues the error message:
**convert-im6.q16: non-conforming drawing primitive definition `color' @ error/draw.c/RenderMVGContent/4404.**
Thank you for your efforts.

**Reply by fmw42** [reply]
It is not the color that it is objecting, but that IM 6 does not permit inline computations such as I used in the second example, which is for IM 7 using magick.

**Comment by fmw42** [comment, marked as answer]
I believe what you want is the following. You had it almost. You needed to turn off the alpha channel before the filltoborder and then add -transparent white.

`convert yellow_hexagon.png -alpha off -fuzz 10% -fill none -bordercolor yellow -draw "matte 1,1 filltoborder" -transparent white x.png`

![x](https://user-images.githubusercontent.com/13458218/181682636-56d7ece4-ea40-469d-b4f8-890a7ff30578.png)

**Reply by dakra137** [reply]
I think your succinct solution should go into the unfinished  https://imagemagick.org/Usage/masking/#border_cut 
Thank you.
